discoverd/client/client.go

```diff
@@ -36,268 +26,256 @@ func init() {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.leader, c.pinned
 }
 
-func (c *Client) Service(name string) Service {
-	return newService(c, name)
+// Update the currently pinned server
+func (c *Client) updatePin(new int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.pinned = new
 }
 
-func IsNotFound(err error) bool {
-	return hh.IsObjectNotFoundError(err)
+func (c *Client) updateServers(servers []string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.servers = make([]*httpclient.Client, 0, len(servers))
+	for _, s := range formatURLs(servers) {
+		c.servers = append(c.servers, c.httpClient(s))
+	}
 }
 
-func (c *Client) Instances(service string, timeout time.Duration) ([]*Instance, error) {
-	s := c.Service(service)
-	instances, err := s.Instances()
-	if len(instances) > 0 || err != nil && !IsNotFound(err) {
-		return instances, err
+func (c *Client) Do(method string, path string, in, out interface{}, streamReq bool) (res stream.Stream, err error) {
+	var leaderReq bool
+	switch method {
+	case "PUT", "DEL", "POST":
+		leaderReq = true
 	}
 
-	events := make(chan *Event)
-	stream, err := s.Watch(events)
-	if err != nil {
-		return nil, err
-	}
-	defer stream.Close()
-	// get any current instances
-outer:
-	for event := range events {
-		switch event.Kind {
-		case EventKindCurrent:
-			break outer
-		case EventKindUp:
-			instances = append(instances, event.Instance)
-		}
-	}
-	if len(instances) > 0 {
-		return instances, nil
+	leader, pinned := c.currentPin()
+
+	// try direct writes directly to the leader to avoid redirect
+	if leaderReq {
+		pinned = leader
 	}
-	// wait for an instance to come up
-	for {
-		select {
-		case event, ok := <-events:
-			if !ok {
-				return nil, stream.Err()
+
+	c.mu.RLock()
+	servers := make([]*httpclient.Client, len(c.servers))
+	copy(servers, c.servers)
+	c.mu.RUnlock()
+
+	errors := make([]string, 0, len(servers))
+	for i := pinned; i < len(servers)+pinned; i++ {
+		k := i % len(servers)
+		hc := c.servers[k]
+		var rsp *http.Response
+		if streamReq {
+			h := http.Header{"Accept": []string{"text/event-stream"}}
+			rsp, err = hc.RawReq(method, path, h, in, nil)
+			if err == nil {
+				res = httpclient.Stream(rsp, out)
 			}
-			if event.Kind != EventKindUp {
-				continue
+		} else {
+			h := http.Header{"Accept": []string{"application/json"}}
+			rsp, err = hc.RawReq(method, path, h, in, out)
+			if err == nil && out == nil {
+				rsp.Body.Close()
 			}
-			return []*Instance{event.Instance}, nil
-		case <-time.After(timeout):
-			return nil, ErrTimedOut
 		}
+		if err != nil && isNetError(err) {
+			errors = append(errors, err.Error())
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+		// If the pinned server failed to fufil our request then update the pin
+		// We don't update the pin on leader requests
+		if k != pinned && !leaderReq {
+			c.updatePin(k)
+		}
+		if peers := rsp.Header.Get("DISCOVERD_PEERS"); peers != "" {
+			c.updateServers(strings.Split(peers, ","))
```

Comment:
Should this include the raft index so that out of order request response processing doesn't nuke a new peer list?